### PR TITLE
2910 and 2446: Fixes to auto-restart after preferences change

### LINKED ISSUE #2446: `print` statements fail after preference restart
**Describe the bug**

`OSError(22)` is produced by print statements after an automatic restart is performed.

**To Reproduce**
Run `run.py` (I used PyCharm)
Change a preference that requires restart
Save, and run some code that has a print statement

**SasView version (please complete the following information):**
 - Version: main

### PR DESCRIPTION
**Pull request by krzywon**
## Description

This fixes the issues related to restarting the installed version of SasView after making a preference change. This allows for different behaviors when running from python or from the executable.

Fixes #2446, #2910

## How Has This Been Tested?

I tested the installed version from https://github.com/SasView/sasview/actions/runs/9893932935 for both Windows and Mac. The auto-restart seems to work when changing the DPI and scaling values.

## Review Checklist:

**Documentation** (check at least one)
- [x] There is **nothing** that needs documenting
- [ ] Documentation changes are **in this PR**
- [ ] There is an **issue** open for the documentation (link?)

**Installers**
- [x] There is a chance this will affect the **installers**, if so
  - [ ] **Windows** installer (GH artifact) has been tested (installed and worked) 
  - [ ] **MacOSX** installer (GH artifact) has been tested (installed and worked) 

**Licencing** (untick if necessary)
- [x] The introduced changes comply with SasView license (BSD 3-Clause)

### CODE REVIEW

**Review by wpotrzebowski — COMMENTED** [review]
It seems that there are only two options i.e. python and main executable do we need "pass" 3 times?

**Comment by krzywon** [comment]
> do we need "pass" 3 times?

Whoops, no. I had those in for debugging and just removed them.

File: src/sas/qtgui/Utilities/Preferences/PreferencesPanel.py [hunk 1]
```diff
@@ -30,6 +30,19 @@
...
 
 
+def restart_main():
+    # Restart the application...
+    executable = sys.executable
+    executable_filename = os.path.split(executable)[1]
+    if executable_filename.lower().startswith('python'):
+        # application is running within a python interpreter
+        python = executable
+        os.execv(python, [python, ] + sys.argv)
```
Comment (lucas-wilkins) [inline src/sas/qtgui/Utilities/Preferences/PreferencesPanel.py:40]: This line looks weird, are we sure its right? Should one be in quotes, just checking

**Review by lucas-wilkins — APPROVED** [review]
Code looks mainly sensible, bit unsure about one line though. Going to approve, but please check

File: src/sas/qtgui/Utilities/Preferences/PreferencesPanel.py (hunk 1 above)
Comment (krzywon) [reply src/sas/qtgui/Utilities/Preferences/PreferencesPanel.py:40]: Both pythons are already strings, so no quotes needed, but the extra one is required for the restart when running from source.

**Review by butlerpd — APPROVED** [review]
tested windows installer and works as advertised on re-start.  That said, I still worry about those changes that require a restart. I used scale = 0.5 and while it restarted it was not only unusable I could not even get to the preferences menu to fix. I had to edit the config-6.json file.   We probably need another ticket for that. Something like a restart popup that is NOT scaled asking "keep changes or reject changes" maybe?

That however is not related to this PR so will also approve on functionality so ready to merge

**Comment by lucas-wilkins** [comment]
> I could not even get to the preferences menu to fix

What do you mean by that?

**Comment by krzywon** [comment]
> > I could not even get to the preferences menu to fix
> 
> What do you mean by that?

I'm guessing the scaling was so bad that he couldn't access the menu to open the preferences panel. This is a potential real issue and needs to be fixed - see #2388.

**Pull request merged**